src/integration-test/java/com/commercetools/sync/integration/services/impl/ShoppingListServiceImplIT.java

```diff
@@ -0,0 +1,265 @@
+    private List<String> errorCallBackMessages;
+    private List<Throwable> errorCallBackExceptions;
+
+    /**
+     * Deletes shopping list and products from the target CTP projects, then it populates the project with test data.
+     */
+    @BeforeEach
+    void setup() {
+        deleteShoppingListTestData(CTP_TARGET_CLIENT);
+        errorCallBackMessages = new ArrayList<>();
+        errorCallBackExceptions = new ArrayList<>();
+        shoppingList = createShoppingList(CTP_TARGET_CLIENT, "name", "key");
+        final ShoppingListSyncOptions options =
+            ShoppingListSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+                                          .errorCallback((exception, oldResource, newResource, actions) -> {
+                                              errorCallBackMessages.add(exception.getMessage());
+                                              errorCallBackExceptions.add(exception);
+                                          })
+                                          .build();
+        shoppingListService = new ShoppingListServiceImpl(options);
+    }
+
+    /**
+     * Cleans up the target test data that were built in this test class.
+     */
+    @AfterAll
+    static void tearDown() {
+        deleteShoppingLists(CTP_TARGET_CLIENT);
+    }
+
+    @Test
+    void fetchShoppingList_WithNonExistingShoppingList_ShouldReturnEmptyOptional() {
+        final Optional<ShoppingList> shoppingList =
+            shoppingListService.fetchShoppingList("not-existing-key").toCompletableFuture().join();
+
+        assertThat(shoppingList).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+
+    }
+
+    @Test
+    void fetchShoppingList_WithExistingShoppingList_ShouldFetchShoppingList() {
+        final Optional<ShoppingList> shoppingList =
+            shoppingListService.fetchShoppingList(this.shoppingList.getKey()).toCompletableFuture().join();
+
+        assertThat(shoppingList).isNotEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchMatchingShoppingListsByKeys_WithNotExistingKeys_ShouldReturnEmptySet() {
+        final Set<String> shoppingListKeys = new HashSet<>();
+        shoppingListKeys.add("not_existing_key_1");
+        shoppingListKeys.add("not_existing_key_2");
+        Set<ShoppingList> shoppingLists =
+            shoppingListService.fetchMatchingShoppingListsByKeys(shoppingListKeys).toCompletableFuture().join();
+
+        assertThat(shoppingLists).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+    }
+
+    @Test
+    void fetchMatchingShoppingListsByKeys_WithExistingShoppingListsKeys_ShouldReturnShoppingLists() {
+        Set<ShoppingList> shoppingLists =
+            shoppingListService.fetchMatchingShoppingListsByKeys(singleton(shoppingList.getKey()))
```

Comment:
normally, I prefer 2 objects at least for this method to ensure the query is working as expected 🤔

Reply:
2365da1